API Documentation (#44)
rapidsai/cuvs: docs/source/build.md

**# Installation

> The cuVS software development kit provides APIs for C, C++, Python, and Rust languages. These 

Reviewer: Unfinished sentence?

> The cuVS software development kit provides APIs for C, C++, Python, and Rust languages. These 

Reviewer 2: Trailing sentence here - maybe remove ? 
```suggestion
The cuVS software development kit provides APIs for C, C++, Python, and Rust languages. 
```

Reviewer 2: I see @dantegd has made the same comment - nvm

    - [Build dependencies](#build-dependencies)
      - [Required](#required)
  - [Create a build environment](#creating-a-build-environment)
      - [ccache and sccache](#ccache-and-sccache)
  - [C/C++ Libraries](#c-library)
      - [Building the Googletests](#building-the-googletests)
  - [Python Library](#python-library)
  - [Rust Library](#rust-library)
  - [Using Cmake Directly](#using-cmake-directly)
- [Build Documentation](#build-documentation)
- [Use cuVS in your application](#use-cuvs-in-your-application)

[//]: # (- [Using cuVS in downstream projects]&#40;#using-raft-c-in-downstream-projects&#41;)

[//]: # (  - [CMake targets]&#40;#cmake-targets&#41;_)

------

## Installing Pre-compiled Packages

### C, C++, and Python through Conda

The easiest way to install the pre-compiled C, C++, and Python packages is through [conda](https://docs.anaconda.com/free/miniconda/index.html). Use the following commands, depending on your CUDA version, to install cuVS packages (replace `rapidsai` with `rapidsai-nightly` to install more up-to-date but less stable nightly packages). `mamba` is preferred over the `conda` command.

#### C++ Package
```bash

> mamba install -c rapidsai -c conda-forge -c nvidia libcuvs cuda-version=11.8

Reviewer: ```suggestion
mamba install -c rapidsai -c conda-forge -c nvidia libcuvs cuda-version=12.0
```

I think making 12.0 the default example/suggestion is appropriate at this time

```

#### C Package
```bash

> mamba install -c rapidsai -c conda-forge -c nvidia libcuvs_c cuda-version=11.8

Reviewer: ```suggestion
mamba install -c rapidsai -c conda-forge -c nvidia libcuvs_c cuda-version=12.0
```

same as above

```

#### Python Package
```bash

> mamba install -c rapidsai -c conda-forge -c nvidia pycuvs cuda-version=11.8

Reviewer: ```suggestion
mamba install -c rapidsai -c conda-forge -c nvidia pycuvs cuda-version=12.0
```

```

### Python through Pip

The cuVS Python package can also be [installed through pip](https://rapids.ai/pip.html#install). 

For CUDA 11 packages:
```bash
pip install pycuvs-cu11 --extra-index-url=https://pypi.nvidia.com
```

And CUDA 12 packages:
```bash
pip install pycuvs-cu12 --extra-index-url=https://pypi.nvidia.com
```

> Note: these packages statically links the C and C++ libraries so the `libcuvs` and `libcuvs_c` shared libraries won't be readily available to use in your code. 

Reviewer: ```suggestion
Note: these packages statically link the C and C++ libraries so the `libcuvs` and `libcuvs_c` shared libraries won't be readily available to use in your code. 
```

### Prerequisites

- Cmake 3.26.4+
- GCC 9.3+ (11.4+ recommended)
- CUDA Toolkit 11.4+
- Volta architecture or better (compute capability >= 7.0)

### Create a build environment

[Conda](https://docs.anaconda.com/free/miniconda/index.html) environment scripts are provided for installing the necessary dependencies to build cuVS from source. It is preferred to use `mamba`, as it provides significant speedup over `conda`:
```bash
mamba env create --name cuvs -f conda/environments/all_cuda-120_arch-x86_64.yaml
mamba activate cuvs
```

The process for building from source with CUDA 11 differs slightly in that your host system will also need to have CUDA toolkit installed which is greater than, or equal to, the version you install into you conda environment. Installing CUDA toolkit into your host system is necessary because `nvcc` is not provided with Conda's cudatoolkit dependencies for CUDA 11. The following example will install create and install dependencies for a CUDA 11.8 conda environment
```bash
mamba env create --name cuvs -f conda/environments/all_cuda-118_arch-x86_64.yaml
mamba activate cuvs
```

> The recommended way to build and install cuVS from source is to use the `build.sh` script in the root of the repository. This script can build both the C++ and Python artifacts and provides CMake options for building and installing the headers, tests, benchmarks, and the pre-compiled shared library.

Reviewer: More of a curiosity question probably for @benfred than anything else, can we add the rust build also to build.sh? Should be possible since it's just doing the `cargo` command, no?

Reviewer 2: I'll be adding the rust build to build.sh in a future PR - we can definitely add this in

Author: I was talking w/ Bed about that earlier. Yes, the plan is ultimately to add that to build.sh

### C and C++ libraries

The C and C++ shared libraries are built together using the following arguments to `build.sh`:
```bash
./build.sh libcuvs
```

In above example the `libcuvs.so` and `libcuvs_c.so` shared libraries are installed by default into `$INSTALL_PREFIX/lib`. To disable this, pass `-n` flag.

Once installed, the shared libraries, headers (and any dependencies downloaded and installed via `rapids-cmake`) can be uninstalled using `build.sh`:
```bash
./build.sh libcuvs --uninstall
```

#### Building the Googletests

Compile the C and C++ Googletests using the `tests` target in `build.sh`.

```bash
./build.sh libcuvs tests
```

The tests will be written to the build directory, which is `cpp/build/` by default, and they will be named `*_TEST`.

> It can take sometime to compile all of the tests. You can build individual tests by providing a semicolon-separated list to the `--limit-tests` option in `build.sh`. Make sure to pass the `-n` flag so the tests are not installed.

Reviewer: ```suggestion
It can take some time to compile all of the tests. You can build individual tests by providing a semicolon-separated list to the `--limit-tests` option in `build.sh`. Make sure to pass the `-n` flag so the tests are not installed.
```

```bash

> ./build.sh libcuvs tests -n --limit-tests=NEIGHBORS_TEST;CLUSTER_TEST

Reviewer: ```suggestion
./build.sh libcuvs tests -n --limit-tests= NEIGHBORS_ANN_CAGRA_TEST; CAGRA_C_TEST
```

Using examples of tests we've already checked in might be a good idea to avoid initial potential confusion for a user
